index.js

```diff
@@ -60,43 +17,95 @@ module.exports = function(opts, data, options){
 
   var template;
 
-  if (typeof(opts) === 'object') {
-    template = fs.readFileSync(opts.src, 'utf-8');
+  if (typeof opts === 'object') {
+    if (typeof opts.src !== 'string') {
+      throw new PluginError(PLUGIN_NAME, 'Expecting `src` option.');
+    }
+    template = fs.readFileSync(opts.src, 'utf8');
   } else {
     template = opts;
   }
+  
+  data = data || {};
+  options = options || {};
+
+  if (!options.engine) {
+    options.engine = 'lodash';
+  }
 
-  function wrap(file, callback){
-    if (gutil.isStream(file.contents)) {
-      var through = es.through();
-      var wait = es.wait(function(err, contents){
-        compile(file, contents, template, data, options, function(compileErr, output) {
-          if (compileErr) {
-            callback(compileErr);
-          } else {
-            through.write(output);
-            through.end();
+  return through.obj(function(file, enc, cb){
+    if (file.isNull()) {
+      cb(null, file);
+      return;
+    }
+
+    function compile(contents, done){
+      // attempt to parse the file contents for JSON or YAML files
+      if (options.parse !== false) {
+        try {
+          if (file.path.match(/json$/)) {
+            contents = JSON.parse(contents);
+          } else if (file.path.match(/ya?ml$/)) {
+            contents = require('js-yaml').safeLoad(contents);
           }
-        });
-      });
+        } catch (err) {
+          throw new PluginError(PLUGIN_NAME, PLUGIN_NAME + ': error parsing ' + file.path);
+        }
+      }
+
+      /*
+       * Add `file` field to source obj used when interpolating
+       * the template. Ensure the user supplied data object is not
+       * augmented to prevent file prop leakage across multiple
+       * streams. Properties specified in the user supplied data
+       * object should take precedence over properties supplied
+       * by the file.
+       */
+      var newData = extend({ file: file }, file.data, data, options, { contents: contents });
+
+      /*
+       * Allow template to be a function, pass it the data object.
+       */
+      if (typeof template === 'function') {
+        template = template(newData);
+      }
 
-      file.contents.pipe(wait);
-      file.contents = through;
-      callback(null, file);
-    } else if (gutil.isBuffer(file.contents)) {
-      compile(file, file.contents.toString('utf-8'), template, data, options, function(compileErr, output) {
-        if (compileErr) {
-          callback(compileErr);
-        } else {
-          file.contents = new Buffer(output);
-          callback(null, file);
+      consolidate[options.engine].render(template, newData, function(err, result) {
+        if (err) {
+          done(new PluginError(PLUGIN_NAME, err));
+          return;
         }
+
+        done(null, new Buffer(result));
       });
-    } else {
-      callback(null, file);
     }
 
-  }
+    var self = this;
```

Comment:
You might consider putting the `var` at the top of the function to mirror where it is hoisted too.